File: Go/Go_1000-/1331.go
// Go
// leetcode 1331
// https://leetcode.cn/problems/rank-transform-of-an-array/

func arrayRankTransform(arr []int) []int {
	n := len(arr)
	res := make([]int, n)
	copy(res, arr)
	sort.Ints(arr)

	// 把排名录入 map m中
	m := make(map[int]int, n)
	idx := 1
	for _, v := range arr {
		if _, ok := m[v]; ok {
			continue
		}

		m[v] = idx
		idx++
	}

	// 填入答案
	for i := 0; i < n; i++ {
		res[i] = m[res[i]]
	}

	return res
}

File: cpp/cpp_1000-/1331.cc
// C++
// leetcode 1331
// https://leetcode.cn/problems/rank-transform-of-an-array/

class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
    	int n = arr.size();
    	map<int, int> m;

    	// 录入数据
    	for (auto x : arr)
    		m[x] = 0;

    	// 排名
        int p = 1;
        for (auto &[k ,v] : m)
        	v = p++;

        // 填装答案
    	for (int i = 0; i < n; i++)
    		arr[i] = m[arr[i]];

    	return arr;
    }
};